app/instances/views.py

```diff
@@ -0,0 +1,142 @@
+from flask import render_template, current_app
+from flask_wtf.csrf import generate_csrf
+from flask_login import current_user, login_required
+from urllib.parse import quote
+
+from . import instances
+from ..utils import get_heroku_token, register_subdomain
+from .forms import LaunchInstanceForm
+from ..models import Instance
+from ..decorators import heroku_auth_required
+from .. import db
+
+import string
+import random
+import requests
+import re
+
+
+@instances.route('/heroku-authorize')
+@login_required
+def heroku_authorize():
+    link = 'https://id.heroku.com/oauth/authorize?' +\
+           'client_id={}&response_type=code&scope={}&state={}'.format(
+            current_app.config['HEROKU_CLIENT_ID'], 'global',
+            quote(generate_csrf()))
+    return render_template('instances/heroku_authorize.html', oauth_link=link)
+
+
+def generate_secret(pass_len):
+    s = string.ascii_letters + string.digits
+    p = ''.join(random.SystemRandom().choice(s) for _ in range(pass_len))
+    return p
+
+
+@instances.route('/verify-needed', methods=['GET'])
+def require_verification():
+    return render_template('instances/verify_needed.html')
+
+
+@instances.route('/launch', methods=['GET', 'POST'])
+@login_required
+@heroku_auth_required
+def launch():
+    form = LaunchInstanceForm()
+    if form.validate_on_submit():
+        name = form.name.data
+
+        username_in_app = current_user.email
+        password_in_app = generate_secret(8)
+
+        with requests.Session() as s:
+            s.trust_env = False
+            auth = get_heroku_token(
+                s,
+                current_user.heroku_refresh_token,
+                current_app.config['HEROKU_CLIENT_SECRET']
+            )
+
+            headers = {
+                'Authorization': 'Bearer {}'.format(auth),
+                'Accept': 'application/vnd.heroku+json; version=3'
+            }
+
+            data = {
+                'source_blob': {
+                    'url': 'https://github.com/hack4impact/maps4all/tarball/master'  # noqa
+                },
+                'overrides': {
+                    'env': {
+                        # TODO: Doesn't this effectively leak our Twilio and
+                        # Sendgrid accounts and put more load on our accounts?
```

Comment:
as mentioned before i dont think there's anything we could do about twilio. but there's an addon for sendgrid. https://elements.heroku.com/addons/sendgrid